### PR TITLE
feat(tls_codec): Implement `tls_codec(skip)` support.

### PR DESCRIPTION
**Pull request by duesee**
This PR implements support for a `skip` attribute on structs. When using ...

```rust
#[derive(TlsDeserialize, TlsSerialize, TlsSize)]
struct StructWithSkip {
  #[tls_codec(skip)]
  a: u8,
  b: u8,
}
```

... only `b` will be serialized. During deserialization `Default::default()` is called to populate `a`.

When `Default` is not implemented for an attribute we get something like ...

```
error[E0277]: the trait bound `NonDefaultField: Default` is not satisfied
  --> tls_codec/derive/tests/attributes.rs:34:14
   |
34 |     #[derive(TlsDeserialize, TlsSerialize, TlsSize)]
   |              ^^^^^^^^^^^^^^ the trait `Default` is not implemented for `NonDefaultField`
```

It is also not allowed to use `skip` twice on the same attribute. Open question: should we accept `skip` case-insensitive?

This is a first (unpolished) working version but it may have some problems I overlooked. There is also no documentation yet. First time doing something with procedural macros :-)

### CODE REVIEW

File: tls_codec/derive/src/lib.rs [hunk 1]
```diff
@@ -151,13 +152,20 @@ enum TlsAttr {
...
 }
 
 impl TlsAttr {
     fn name(&self) -> &'static str {
         match self {
             TlsAttr::With(_) => "with",
             TlsAttr::Discriminant(_) => "discriminant",
+            TlsAttr::Skip => "skip",
         }
     }
```
Comment (franziskuskiefer) [inline tls_codec/derive/src/lib.rs:171]: Can you update the comment on `parse`?

File: tls_codec/derive/src/lib.rs [hunk 2]
```diff
@@ -232,16 +254,29 @@ fn function_prefix(field: &Field) -> Result<Prefix> {
...
-                format!("Unrecognized field attribute `{}`", attr.name()),
-            )),
+            (path, _) => Ok(path),
         })?
         .map(Prefix::Custom)
         .unwrap_or_else(|| Prefix::Type(field.ty.clone()));
     Ok(prefix)
 }
 
+fn function_skip(field: &Field) -> Result<bool> {
```
Comment (franziskuskiefer) [inline tls_codec/derive/src/lib.rs:264]: Please add comments to function to exaplain what they are doing.

File: tls_codec/derive/src/lib.rs [hunk 3]
```diff
@@ -424,7 +465,27 @@ fn impl_tls_size(parsed_ast: TlsStruct) -> TokenStream2 {
...
             member_prefixes,
+            member_skips,
         }) => {
+            // Filter out skipped members.
+            let (members, member_prefixes) = {
+                let mut tmp_members: Vec<Member> = Vec::new();
+                let mut tmp_member_prefixes: Vec<Prefix> = Vec::new();
+
+                for ((member, prefix), skip) in members
+                    .into_iter()
```
Comment (franziskuskiefer) [inline tls_codec/derive/src/lib.rs:476]: Using `iter` everywhere here and only cloning the `member` when needed should be more efficient.

File: tls_codec/derive/src/lib.rs [hunk 4]
```diff
@@ -497,7 +558,27 @@ fn impl_serialize(parsed_ast: TlsStruct) -> TokenStream2 {
             generics,
             members,
             member_prefixes,
+            member_skips,
         }) => {
+            // Filter out skipped members.
+            let (members, member_prefixes) = {
+                let mut tmp_members: Vec<Member> = Vec::new();
```
Comment (franziskuskiefer) [inline tls_codec/derive/src/lib.rs:565]: Same code as above? Can we pull this out to avoid code duplication?

File: tls_codec/derive/src/lib.rs [hunk 5]
```diff
@@ -587,7 +668,37 @@ fn impl_deserialize(parsed_ast: TlsStruct) -> TokenStream2 {
             generics,
             members,
             member_prefixes,
+            member_skips,
         }) => {
+            // Partition skipped and non-skipped members.
+            let (members, member_prefixes, members_default, member_prefixes_default) = {
+                let mut tmp_members: Vec<Member> = Vec::new();
```
Comment (franziskuskiefer) [inline tls_codec/derive/src/lib.rs:675]: Almost the same here.

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
Genereally looks good to me with a few comments inline.

**Comment by duesee** [comment]
Thanks for the review. I'll cleanup, deduplicate, and add a bit of documentation.

**Comment by duesee** [comment]
I used the `trybuild` crate. Not sure if this is too much. But it might be good to already have that in place in case we want to work on user-facing errors in the future. The alternative is, as you said, to use doc comments with `compile_fail` on private functions.

File: tls_codec/derive/Cargo.toml [hunk 6]
```diff
@@ -20,3 +20,4 @@ proc-macro2 = "1.0"
 
 [dev-dependencies]
 tls_codec = { path = "../" }
+trybuild = "*"
```
Comment (franziskuskiefer) [inline tls_codec/derive/Cargo.toml:23]: Let's use a version.

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
I think using `trybuild` is fine. Only one nit.

**Review by franziskuskiefer — APPROVED** [review]
Thanks

**Comment by duesee** [comment]
I should have mentioned that I intended to cleanup the commits before merging. It's okay but having "fixups" in the commit history is not great. Sorry for that.

**Comment by franziskuskiefer** [comment]
It's getting squashed anyway 🙂

**Comment by duesee** [comment]
Ah, true that :-)

**Pull request merged**